nexus/src/app/disk.rs

```diff
@@ -404,20 +405,61 @@ impl super::Nexus {
         let _authz_org: authz::Organization;
         let authz_project: authz::Project;
         let authz_disk: authz::Disk;
+        let db_disk: db::model::Disk;
 
-        (authz_silo, _authz_org, authz_project, authz_disk) =
+        (authz_silo, _authz_org, authz_project, authz_disk, db_disk) =
             LookupPath::new(opctx, &self.db_datastore)
                 .organization_name(organization_name)
                 .project_name(project_name)
                 .disk_name(&db::model::Name(params.disk.clone()))
-                .lookup_for(authz::Action::Read)
+                .fetch_for(authz::Action::Read)
                 .await?;
 
+        // If there isn't a running propolis, Nexus needs to use the Crucible
+        // Pantry to make this snapshot
+        let use_the_pantry = if let Some(attach_instance_id) =
```

Comment:
I'm seeing that we make the call on whether or not to use the pantry *super* early - like, before the saga even starts. I think that might run into some issue due to concurrent requests.

The current structure of this code is roughly:

1. Read Disk state from CRDB to see if there is an attached instance ID.
2. Use the output from (1) when constructing the saga DAG.
3. Run the actual saga.

What if the following ordering happens:

1. Read Disk state from CRDB to see if there is an attached instance ID. Let's pretend there *is* an attached disk.
**Let's suppose a concurrent request arrives now, and stops + destroys the instance** 
2. Use the output from (1) when constructing the saga DAG. The structure of the DAG assumes that it can contact the instance, and doesn't use the pantry.
3. Run the actual saga. See that the instance does not exist. Fail?

Would there be a problem with *unconditionally* sending the request to the pantry?

Reply:
I believe the above situation should result in the snapshot create saga failing.
I've added some more return `ActionError` when the snapshot create saga finds things different
than it expects, and I think returning error if things have changed under the snapshot is the
"right" thing to do here at this time.

Reply:
This still seems really sketchy to me. We explicitly throw an error for instances that are in unexpected states (see: `cannot snapshot attached disk for instance in state...`), but we don't really have a great way to validate that these sorts of transitions didn't happen immediately after our check. 

What about the other way? What if we didn't have an instance (so we expected to use the pantry), but we spin one up concurrently between performing this check and doing the saga? Would it be a problem to have the disk attached to the pantry at the same time as it's attached to an instance?

Reply:
I believe I now get what you are worried about here.

Your first `what if` was the right idea, but you picked a path where things would fail gracefully. 
It can indeed fail if:

1: The disk is not attached to an instance at the time of check, (pantry = true) 
  - After that check a new instance is created and the disk is attached to it.
  - Pantry will then steal the disk away from the instance.

2: Disk attached to a stopped instance at the time of check. (pantry = true)
  - After the check, the instance is started.
  - Pantry will steal the disk away from the instance.

I was able to get the pantry (with some well placed sleeps) to steal a disk away from an instance by doing 1 above.

I'm going to have to ponder this morass for a bit.

Ideally, either the snapshot can run to completion, or a change anywhere in the middle will cause the
snapshot to abort and cleanup behind itself.

Reply:
84825d09 adds a check for this disk attach race, let me know what you think

Reply:
> Is it a desirable property that "attaching or detaching a disk to an instance" causes a concurrent snapshot request to fail? I'm trying to wrap my head around "should we be handling the snapshot request gracefully either way, or should we force a failure if this edge condition happens?"

To elaborate from our chat: our current implementation can't support concurrent operations where a disk's volume is being activated by multiple Upstairs, so we have to cause these concurrent requests to fail.

> It seems like the disk still exists in either state, so it really could be snapshottable at any time, right?

From the perspective of an ideal cloud, a disk should be snapshottable at any time, yeah :) We may support this in the future but certainly not today.

> In the case where we want to throw an error, are we propagating the correct error back to the user? We should tell them to retry, but I suspect a string-based ActionError::action_failed error translates to a 500 error.

Yeah, this is likely a 500. I'm not aware of any HTTP response code that fits nicely with "this will work if you retry it but can't work now". The user sent a well-formed request but the server was unable to process it, so 500 kinda fits.

A 503 might fit here? Problem is user code might not interpret a 503 as "please try again" :)

> Are we testing this anywhere?

To test this, we would need machinery in our integration test framework to construct two sagas, and execute individual nodes from those sagas manually.